muxink/src/demux.rs

```diff
@@ -0,0 +1,254 @@
+    }
+}
+
+/// A handle to a demultiplexer.
+///
+/// A handle is bound to a specific channel, see [`Demultiplexer::create_handle`] for details.
+pub struct DemultiplexerHandle<S> {
+    /// Which channel this handle is listening on
+    channel: u8,
+    /// A reference to the underlying demultiplexer.
+    demux: Arc<Mutex<Demultiplexer<S>>>, // (probably?) make sure this is a stdmutex
+}
+
+impl<S> Stream for DemultiplexerHandle<S>
+where
+    S: Stream<Item = Bytes> + Unpin,
+{
+    // TODO Result<Bytes, Error>
+    type Item = Bytes;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Lock the demultiplexer
+        let mut demux = match self.demux.as_ref().try_lock() {
+            Err(_err) => panic!("TODO"), // TODO return Err("TODO")
+            Ok(guard) => guard,
+        };
+
+        // If next_frame has a suitable frame for this channel, return it in a Poll::Ready. If it
+        // has an unsuitable frame, return Poll::Pending. Otherwise, we attempt to read from
+        // the stream.
+        if let Some((ref channel, ref bytes)) = demux.next_frame {
+            if *channel == self.channel {
+                let bytes = bytes.clone();
+                demux.next_frame = None;
+                return Poll::Ready(Some(bytes));
+            } else {
+                return Poll::Pending;
+            }
+        }
+
+        // Try to read from the stream, placing the frame into next_frame and returning
+        // Poll::Pending if its in the wrong channel, otherwise returning it in a Poll::Ready.
+        match demux.stream.poll_next_unpin(cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(Some(bytes)) => {
+                let channel: u8 = bytes[0..1][0];
+                let frame = bytes.slice(1..);
+                if channel == self.channel {
+                    Poll::Ready(Some(frame))
+                } else {
+                    demux.next_frame = Some((channel, frame));
+                    Poll::Pending
+                }
+            }
+            Poll::Ready(None) => Poll::Ready(None),
+        }
+
+        // TODO: figure out when we are being polled again, does it work correctly (see waker) or
+        //       will it cause inefficient races? do we need to call wake? probably. (possibly
+        //       necessary) can have table of wakers to only wake the right one.
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::marker::Unpin;
+
+    use super::*;
+    use futures::{FutureExt, Stream, StreamExt};
+
+    // This stream is used because it is not safe to call it after it returns
+    // [`Poll::Ready(None)`], whereas many other streams are. The interface for
+    // streams says that in general it is not safe, so it is important to test
+    // using a stream which has this property as well.
+    struct TestStream<T> {
+        // The items which will be returned by the stream in reverse order
+        items: Vec<T>,
```

Comment:
We could use a `VecDeque` instead, that will allow us to add items to it after construction, should that feature be necessary in later tests. We could simulate a temporarily blocked stream that way (either with an explicit `finish()` method or by storing `Option`s of items). Just a thought at this point though.